File: proxy/backfill.js
const db = require("./db");

let aborted = false;
let running = false;

function abort() {
  aborted = true;
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function isoDate(d) {
  return d.toISOString().slice(0, 10);
}

function enumerateDates(start, end) {
  const out = [];
  const d = new Date(Date.UTC(
    start.getUTCFullYear(), start.getUTCMonth(), start.getUTCDate()
  ));
  const stop = Date.UTC(end.getUTCFullYear(), end.getUTCMonth(), end.getUTCDate());
  while (d.getTime() <= stop) {
    out.push(isoDate(d));
    d.setUTCDate(d.getUTCDate() + 1);
  }
  return out;
}

async function runBackfill({
  years = 3,
  fetchOne,
  baseDelayMs = 2000,
  jitterMs = 750,
  log = console.log,
} = {}) {
  if (running) {
    log("backfill: already running, skipping");
    return;
  }
  if (!db.isEnabled()) {
    log("backfill: DB not configured, skipping");
    return;
  }
  if (typeof fetchOne !== "function") {
    throw new Error("backfill: fetchOne(date) required");
  }
  running = true;
  aborted = false;

  try {
    const today = new Date();
    const yesterday = new Date(today); yesterday.setUTCDate(yesterday.getUTCDate() - 1);
    const start = new Date(today); start.setUTCFullYear(start.getUTCFullYear() - years);

    const startStr = isoDate(start);
    const endStr = isoDate(yesterday);
    const scraped = await db.getDatesWithScrape(startStr, endStr);
    const all = enumerateDates(start, yesterday);
    const missing = all.filter((d) => !scraped.has(d));

    log(`backfill: ${missing.length}/${all.length} dates to fetch (${startStr} → ${endStr})`);
    if (!missing.length) return;

    let consecutiveFailures = 0;
    let processed = 0;
    const startedAt = Date.now();

    for (const date of missing) {
      if (aborted) {
        log(`backfill: aborted after ${processed} dates`);
        return;
      }
      try {
        await fetchOne(date);
        consecutiveFailures = 0;
      } catch (err) {
        consecutiveFailures += 1;
        log(`backfill ${date} failed: ${err.message}`);
        if (consecutiveFailures >= 20) {
          log(`backfill: aborting after ${consecutiveFailures} consecutive failures (DOFbasen blocking?)`);
          return;
        }
        if (consecutiveFailures >= 5) {
          const backoffMs = Math.min(60000, 5000 * 2 ** (consecutiveFailures - 5));
          log(`backfill: backing off ${Math.round(backoffMs / 1000)}s after ${consecutiveFailures} consecutive failures`);
          await sleep(backoffMs);
        }
      }
      processed += 1;
      if (processed % 50 === 0) {
        const elapsedSec = Math.round((Date.now() - startedAt) / 1000);
        log(`backfill: ${processed}/${missing.length} done (${elapsedSec}s elapsed)`);
      }
      await sleep(baseDelayMs + Math.random() * jitterMs);
    }

    log(`backfill: complete — processed ${processed} dates in ${Math.round((Date.now() - startedAt) / 1000)}s`);
  } finally {
    running = false;
  }
}

module.exports = { runBackfill, abort };

File: server-rs/src/backfill.rs
//! Historical observation seeder — port of `proxy/backfill.js`. Walks the last
//! N years of dates not already scraped and fetches each with polite delays and
//! consecutive-failure backoff.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use chrono::{Datelike, NaiveDate};

use crate::push::refresh_observations_for_date;
use crate::AppState;

const BASE_DELAY_MS: u64 = 2000;
const JITTER_MS: u64 = 750;

fn enumerate_dates(start: NaiveDate, end: NaiveDate) -> Vec<String> {
    let mut out = Vec::new();
    let mut d = start;
    while d <= end {
        out.push(d.format("%Y-%m-%d").to_string());
        d += chrono::Duration::days(1);
    }
    out
}

/// Small deterministic-ish jitter without pulling in a RNG crate.
fn jitter() -> u64 {
    let nanos = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|d| d.subsec_nanos()).unwrap_or(0);
    (nanos as u64) % JITTER_MS
}

pub async fn run_backfill(st: AppState, years: i32, abort: Arc<AtomicBool>) {
    let Some(db) = st.db.clone() else {
        tracing::info!("backfill: DB not configured, skipping");
        return;
    };

    let today = chrono::Utc::now().date_naive();
    let yesterday = today - chrono::Duration::days(1);
    // year - N, same month/day (clamp Feb 29 → Feb 28).
    let start = NaiveDate::from_ymd_opt(today.year() - years, today.month(), today.day())
        .unwrap_or_else(|| NaiveDate::from_ymd_opt(today.year() - years, today.month(), 28).unwrap());

    let start_str = start.format("%Y-%m-%d").to_string();
    let end_str = yesterday.format("%Y-%m-%d").to_string();

    let scraped = match db.get_dates_with_scrape(&start_str, &end_str).await {
        Ok(s) => s,
        Err(e) => {
            tracing::error!("backfill: scrape-log read failed: {e}");
            return;
        }
    };
    let all = enumerate_dates(start, yesterday);
    let missing: Vec<String> = all.iter().filter(|d| !scraped.contains(*d)).cloned().collect();

    tracing::info!("backfill: {}/{} dates to fetch ({start_str} → {end_str})", missing.len(), all.len());
    if missing.is_empty() {
        return;
    }

    let mut consecutive_failures = 0u32;
    let mut processed = 0usize;

    for date in &missing {
        if abort.load(Ordering::Relaxed) {
            tracing::info!("backfill: aborted after {processed} dates");
            return;
        }
        match refresh_observations_for_date(&st, Some(date.clone())).await {
            Ok(_) => consecutive_failures = 0,
            Err(e) => {
                consecutive_failures += 1;
                tracing::warn!("backfill {date} failed: {e}");
                if consecutive_failures >= 20 {
                    tracing::warn!("backfill: aborting after {consecutive_failures} consecutive failures (DOFbasen blocking?)");
                    return;
                }
                if consecutive_failures >= 5 {
                    let backoff = (5000u64 * 2u64.pow(consecutive_failures - 5)).min(60000);
                    tracing::warn!("backfill: backing off {}s after {consecutive_failures} consecutive failures", backoff / 1000);
                    tokio::time::sleep(Duration::from_millis(backoff)).await;
                }
            }
        }
        processed += 1;
        if processed % 50 == 0 {
            tracing::info!("backfill: {processed}/{} done", missing.len());
        }
        tokio::time::sleep(Duration::from_millis(BASE_DELAY_MS + jitter())).await;
    }

    tracing::info!("backfill: complete — processed {processed} dates");
}
